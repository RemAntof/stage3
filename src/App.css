.appContainer {
  display: flex;
  width: 100%;
  overflow: hidden;
  position: relative;
}

.mainPage {
  flex-shrink: 0;
  width: 100%;
  transition:
    transform 0.3s ease-in-out,
    width 0.3s ease-in-out;
}

#outlet {
  position: fixed;
  top: 0;
  right: 0;
  width: 0;
  height: 100%;
  transition: width 0.3s ease-in-out;
  background-color: #242424;
}

#outlet.visible {
  width: fit-content;
}

.mainPage.shifted {
  width: 60%;
}
