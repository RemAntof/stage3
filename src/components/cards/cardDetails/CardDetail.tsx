import React, { useState, useEffect, useRef } from 'react';
import { useParams } from 'react-router-dom';
import fetchApi from '@services/API/fetchApi';
import Loader from '@components/loader/loader';
import Card from '../card/card';
import Animal from '@interfaces/animal';

const CardDetail: React.FC = () => {
  const { animalName } = useParams<{
    animalName: string;
  }>();
  const [animal, setAnimal] = useState<Animal | null>(null);
  const [loading, setLoading] = useState<boolean>(true);
  const [error, setError] = useState<string | null>(null);
  const cardDetailRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const fetchAnimal = async () => {
      setLoading(true);
      setError(null);
      try {
        const responseData = await fetchApi(animalName);
        setAnimal(responseData.animals[0]);
      } catch (error) {
        setError(error.message);
      } finally {
        setLoading(false);
      }
    };

    fetchAnimal();
  }, [animalName]);

  if (loading) return <Loader />;

  if (error) return <div>Error: {error}</div>;

  return (
    <div ref={cardDetailRef}>
      {animal ? (
        <Card animal={animal} />
      ) : (
        <div>Animal not found</div>
      )}
    </div>
  );
};

export default CardDetail;
