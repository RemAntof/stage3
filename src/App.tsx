import React, { useEffect, useState, useRef } from 'react';
import './App.css';
import MainPage from '@views/mainPage/MainPage';
import {
  Outlet,
  useLocation,
  useNavigate,
} from 'react-router-dom';

const App: React.FC = () => {
  const location = useLocation();
  const [outletVisible, setOutletVisible] = useState(false);
  const outletRef = useRef<HTMLDivElement>(null);
  const navigate = useNavigate();

  useEffect(() => {
    setOutletVisible(location.pathname !== '/');
  }, [location]);

  const onClick = (
    event: React.MouseEvent<HTMLDivElement, MouseEvent>
  ) => {
    if (
      outletVisible &&
      outletRef.current &&
      !outletRef.current.contains(event.target as Node)
    ) {
      setOutletVisible(false);
      navigate(-1);
    }
  };

  return (
    <div className="appContainer" onClick={onClick}>
      <div
        className={`mainPage ${outletVisible ? 'shifted' : ''}`}
      >
        <MainPage />
      </div>
      <div
        id="outlet"
        ref={outletRef}
        className={outletVisible ? 'visible' : 'hidden'}
      >
        {outletVisible ? <Outlet /> : null}
      </div>
    </div>
  );
};

export default App;
