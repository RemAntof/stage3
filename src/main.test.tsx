import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import App from './App';
import ErrorPage from '@views/errorView/errorView';
import CardDetail from '@components/cards/cardDetails/CardDetail';
import { Provider } from 'react-redux';
import store from '@services/store/store';
import {
  createMemoryRouter,
  RouterProvider,
} from 'react-router-dom';

test('renders App component without crashing', () => {
  const router = createMemoryRouter([
    {
      path: '/',
      element: <App />,
      errorElement: <ErrorPage />,
      children: [
        {
          path: ':animalName',
          element: <CardDetail />,
        },
      ],
    },
  ]);

  render(
    <Provider store={store}>
      <RouterProvider router={router} />
    </Provider>
  );

  expect(
    screen.getByTestId('app-container')
  ).toBeInTheDocument();
});
test('matches the snapshot', () => {
  const router = createMemoryRouter([
    {
      path: '/',
      element: <App />,
      errorElement: <ErrorPage />,
      children: [
        {
          path: ':animalName',
          element: <CardDetail />,
        },
      ],
    },
  ]);

  const { asFragment } = render(
    <Provider store={store}>
      <RouterProvider router={router} />
    </Provider>
  );

  expect(asFragment()).toMatchSnapshot();
});
