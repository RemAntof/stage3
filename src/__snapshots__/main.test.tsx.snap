// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="appContainer dark "
    data-testid="app-container"
  >
    <div
      class="mainPage "
    >
      <div
        class="container"
      >
        <input
          id="checkbox"
          type="checkbox"
        />
        <div
          class="background"
        >
          <img
            class="stars"
            src="https://gateway.pinata.cloud/ipfs/Qmcd1rqt5BJrQYX9hHeWc8AqEhBp2pj6Mis82FC5wVW2a2"
          />
          <div
            class="clouds"
          >
            <img
              class="stars"
              src="https://gateway.pinata.cloud/ipfs/QmdDiiwPP4VC5jmRTtPa1v4MrA5JUesq8WZHs8ACLt3Vcj"
            />
            <img
              class="stars"
              src="https://gateway.pinata.cloud/ipfs/QmYNhwZuMNCucpn8YgGxPdKBnVsBDr473pbXCmxKE4uPJa"
            />
          </div>
        </div>
        <label
          class="content"
          for="checkbox"
        >
          <div
            class="sun__moon__container"
          >
            <div
              class="sun__moon"
            >
              <div
                class="sun"
              />
              <div
                class="moon"
              >
                <div />
                <div />
                <div />
              </div>
            </div>
            <div
              class="rays"
            >
              <div />
              <div />
              <div />
            </div>
          </div>
        </label>
      </div>
      <div>
        <h1>
          Something went wrong.
        </h1>
      </div>
    </div>
    <div
      class="hidden"
      data-testid="outlet"
      id="outlet"
    />
  </div>
</DocumentFragment>
`;
